File: atcoder/abc206/d.rs
#![allow(unused)]

fn main() {
    let n = read::<usize>();
    let arr = readv::<usize>();
    let m = *arr.iter().max().unwrap();

    let mut dsu = DSU::new(m + 1);
    for i in 0..n {
        if arr[i] != arr[n - 1 - i] {
            dsu.unite(arr[i], arr[n - 1 - i]);
        }
    }

    let mut ans = 0;
    for x in 1..=m {
        if dsu.root(x) == x && dsu.size(x) >= 2 {
            ans += dsu.size(x) - 1;
        }
    }
    println!("{}", ans);
}

struct DSU {
    par: Vec<usize>,
    siz: Vec<usize>,
}

impl DSU {
    fn new(n: usize) -> Self {
        Self {
            par: (0..n).collect(),
            siz: vec![1; n],
        }
    }

    fn root(&mut self, u: usize) -> usize {
        if self.par[u] == u {
            u
        } else {
            self.par[u] = self.root(self.par[u]);
            self.par[u]
        }
    }

    fn unite(&mut self, mut u: usize, mut v: usize) {
        u = self.root(u);
        v = self.root(v);
        if u == v {
            return;
        }
        if self.siz[u] > self.siz[v] {
            self.par[v] = u;
            self.siz[u] += self.siz[v];
        } else {
            self.par[u] = v;
            self.siz[v] += self.siz[u];
        }
    }

    fn same(&mut self, u: usize, v: usize) -> bool {
        self.root(u) == self.root(v)
    }

    fn size(&mut self, u: usize) -> usize {
        let r = self.root(u);
        self.siz[r]
    }
}

fn read<T: std::str::FromStr>() -> T {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).ok();
    s.trim().parse().ok().unwrap()
}

fn readv<T: std::str::FromStr>() -> Vec<T> {
    read::<String>()
        .split_ascii_whitespace()
        .map(|t| t.parse().ok().unwrap())
        .collect()
}

fn join<T: ToString>(v: &[T], sep: &str) -> String {
    v.iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join(sep)
}

File: atcoder/abc206/d.py
class DisjointSet:
    def __init__(self, N):
        self.N = N
        self.par = list(range(N))
        self.siz = [1 for _ in range(N)]

    def root(self, x):
        if self.par[x] == x:
            return x
        else:
            self.par[x] = self.root(self.par[x])
            return self.par[x]

    def unite(self, x, y):
        x, y = self.root(x), self.root(y)
        if x == y:
            return
        if self.siz[x] > self.siz[y]:
            x, y = y, x
        self.par[x] = y
        self.siz[y] += self.siz[x]

    def same(self, x, y):
        return self.root(x) == self.root(y)

    def size(self, x):
        return self.siz[self.root(x)]


N = int(input())
A = [int(x) - 1 for x in input().split()]
M = max(A)

dsu = DisjointSet(M + 1)
for i in range(N):
    if A[i] != A[N - 1 - i]:
        dsu.unite(A[i], A[N - 1 - i])

ans = 0
for x in range(M + 1):
    if dsu.root(x) == x and dsu.size(x) >= 2:
        ans += dsu.size(x) - 1

print(ans)
